Backporting changes to the piano roll made for MuseScore 4 PR #9109

blackears opened the pull request:
This is a backport of this PR which has been sitting for a long time: #9109

This is a video I put together for the 4.0 PR.  It covers most of what's in this one: https://youtu.be/EwkUV_uXqqc

Basically I implemented a lot of changes requested in an early design document for  the Piano Roll editor.   So you can now:

* middle click & drag to reposition the viewport

* select tool:
** resize the notes by clicking and dragging on their end points
** will work on a group of notes too
* Add tool:
** click and drag to draw in a new note
** click without drag will repeat the length of the last note drawn
*Note length tool:
** Since the functionality of these two are now handled by the select and add tool, it has been repurposed to change playback event times instead.  Adjusting note start and end times will now affect playback timing instead of their notation value

Updated color scheme to separate individual voices.

worldwideweary commented:
Coolio. Out of curiosity and instead of interpreting the code,  what user-experience based changes are being provided here?  A quick glance at the previously linked issues' descriptions didn't seem to have any cogent descriptions or demonstrations of the benefits. A point list would be nice, plus it's good exercise to write that kind of stuff out ;)

blackears commented:
This is a backport of this PR which has been sitting for a long time: https://github.com/musescore/MuseScore/pull/9109

This is a video I put together for the 4.0 OR https://youtu.be/EwkUV_uXqqc

Basically I implemented a lot of changes requested in an early design document for  the Piano Roll editor.   So you can now:

* middle click & drag to reposition the viewport

select tool:
* resize the notes by clicking and dragging on their end points
* will work on a group of notes too
Add tool:
* click and drag to draw in a new note
* click without drag will repeat the length of the last note drawn
Note length tool:
* Since the functionality of these two are now handled by the select and add tool, it has been repurposed to change playback event times instead.  Adjusting note start and end times will now affect playback timing instead of their notation value

Updated color scheme to separate individual voices.

Jojo-Schmitz commented on mscore/pianoroll/pianoroll.cpp:
```diff
@@ -102,14 +102,58 @@ PianorollEditor::PianorollEditor(QWidget* parent)
+          bngrpActionBns->addButton(bn);
+          tbTool->addWidget(bn);
+      }
+
```
Formatting: indenting and bracket style in this whole chunk

Jojo-Schmitz commented:
@wordwideweary: Testers wanted ;-)

blackears commented:
There are some issues where moving notes does not always work cleanly. But this was present in the older version of the PRE too and /i don't know enough about the note engine to be able to resolve this.

Jojo-Schmitz commented on mscore/pianoroll/pianoroll.cpp:
```diff
@@ -102,14 +102,58 @@ PianorollEditor::PianorollEditor(QWidget* parent)
+            PianoRollEditTool _tool;
+            QString _tooltip;
+            bool _selected;
+      };
```
picky mode: brace style

Jojo-Schmitz commented on mscore/pianoroll/pianoroll.cpp:
```diff
@@ -102,14 +102,58 @@ PianorollEditor::PianorollEditor(QWidget* parent)
+            { QStringLiteral(":/data/icons/preEdit-changeLength.svg"), PianoRollEditTool::EVENT_ADJUST, tr("Change Chord Length"), false },
+            { QStringLiteral(":/data/icons/preEdit-tie.svg"), PianoRollEditTool::TIE, tr("Toggle Tie"), false },
+            { "", PianoRollEditTool::LAST, "", false },
+      };
```
dito

Jojo-Schmitz commented on mscore/pianoroll/pianorolledittool.h:
```diff
@@ -17,15 +17,15 @@ namespace Ms {
 
-    LAST  //Marker for end of list - not a tool
+      LAST  //Marker for end of list - not a tool
 };
```
dito

Jojo-Schmitz commented on mscore/pianoroll/pianoview.cpp:
```diff
@@ -478,11 +478,21 @@ void PianoView::drawBackground(QPainter* p, const QRectF& r)
+      p->setRenderHints(QPainter::Antialiasing | QPainter::SmoothPixmapTransform | QPainter::TextAntialiasing);
+      for (PianoItem* block : _noteList) {
+            drawNoteBlock(p, block);
+      }
```
dito

Jojo-Schmitz commented on mscore/pianoroll/pianoview.cpp:
```diff
@@ -510,6 +520,117 @@ void PianoView::drawBackground(QPainter* p, const QRectF& r)
+                  p->drawText(textRect, Qt::AlignLeft | Qt::AlignTop, name);
+                  }
+            }
+}
```
dito

Jojo-Schmitz commented on mscore/pianoroll/pianoview.cpp:
```diff
@@ -609,6 +730,8 @@ void PianoView::wheelEvent(QWheelEvent* event)
       }
 
+
+
```
why?

Jojo-Schmitz commented on mscore/pianoroll/pianoview.cpp:
```diff
@@ -796,18 +954,116 @@ void PianoView::mouseReleaseEvent(QMouseEvent* event)
+//---------------------------------------------------------
+
+void PianoView::finishNoteEventAdjustDrag()
+{
```
brace indent

Jojo-Schmitz commented on mscore/pianoroll/pianoview.cpp:
```diff
@@ -796,18 +954,116 @@ void PianoView::mouseReleaseEvent(QMouseEvent* event)
+            }
+
+      update();
+}
```
dito

Jojo-Schmitz commented on mscore/pianoroll/pianoview.cpp:
```diff
@@ -796,18 +954,116 @@ void PianoView::mouseReleaseEvent(QMouseEvent* event)
+//---------------------------------------------------------
+
+void PianoView::updateCursor()
+{
```
dito

Jojo-Schmitz commented on mscore/pianoroll/pianoview.cpp:
```diff
@@ -796,18 +954,116 @@ void PianoView::mouseReleaseEvent(QMouseEvent* event)
+            }
+
+      setCursor(Qt::ArrowCursor);
+}
```
dito

Jojo-Schmitz commented on mscore/pianoroll/pianoview.cpp:
```diff
@@ -796,18 +954,116 @@ void PianoView::mouseReleaseEvent(QMouseEvent* event)
 
 void PianoView::mouseMoveEvent(QMouseEvent* event)
-      {
+{
```
was better/correct before

Jojo-Schmitz commented on mscore/pianoroll/pianoview.cpp:
```diff
@@ -816,43 +1072,98 @@ void PianoView::mouseMoveEvent(QMouseEvent* event)
+                                    }
+                              else if (_mouseDownPos.x() >= bounds.x() + bounds.width() - _dragNoteLengthMargin) {
+                                    _dragStyle = DragStyle::NOTE_LENGTH_END;
+                              }
```
brace indent

Jojo-Schmitz commented on mscore/pianoroll/pianoview.cpp:
```diff
@@ -816,43 +1072,98 @@ void PianoView::mouseMoveEvent(QMouseEvent* event)
+                              _dragStartPitch = mouseDownPitch;
+                              _dragStartTick = pi->note()->tick();
+                              _dragEndTick = _dragStartTick + pi->note()->chord()->ticks();
+                              _dragNoteCache = serializeSelectedNotes();
```
to much indent

Jojo-Schmitz commented on mscore/pianoroll/pianoview.cpp:
```diff
@@ -1646,9 +1950,9 @@ void PianoView::setStaff(Staff* s, Pos* l)
+      for (Chord* c: chrd->graceNotes())
             addChord(c, voice);
-      for (Note* note : chrd->notes()) {
+      for (Note* note: chrd->notes()) {
```
dito

Jojo-Schmitz commented on mscore/pianoroll/pianoview.cpp:
```diff
@@ -1996,44 +2300,80 @@ void PianoView::pasteNotesAtCursor()
+            Fraction noteStartDraggedAlignedTick = Fraction(noteStartDraggedTick.numerator() * divisions / noteStartDraggedTick.denominator(), divisions);
+            pasteTickOffset = noteStartDraggedAlignedTick - _dragStartTick;
+            pitchOffset = dragToPitch - startPitch;
+      }
```
brace indent style

Jojo-Schmitz commented on mscore/pianoroll/pianoview.cpp:
```diff
@@ -1996,44 +2300,80 @@ void PianoView::pasteNotesAtCursor()
+            Fraction noteEndDraggedTick = _dragEndTick + dragOffsetTicks;
+            Fraction noteEndDraggedAlignedTick = Fraction(noteEndDraggedTick.numerator() * divisions / noteEndDraggedTick.denominator(), divisions);
+            pasteLengthOffset = noteEndDraggedAlignedTick - _dragEndTick;
+      }
```
dito

Jojo-Schmitz commented on mscore/pianoroll/pianoview.cpp:
```diff
@@ -1996,44 +2300,80 @@ void PianoView::pasteNotesAtCursor()
+            Fraction noteStartDraggedAlignedTick = Fraction(noteStartDraggedTick.numerator() * divisions / noteStartDraggedTick.denominator(), divisions);
+            pasteTickOffset = noteStartDraggedAlignedTick - _dragStartTick;
+            pasteLengthOffset = _dragStartTick - noteStartDraggedAlignedTick;
+      }
```
dito

Jojo-Schmitz commented on mscore/pianoroll/pianoview.cpp:
```diff
@@ -1996,44 +2300,80 @@ void PianoView::pasteNotesAtCursor()
-      pasteNotes(_dragNoteCache, pasteTickOffset, pitchOffset, true);
+      if (!(event->modifiers() & Qt::ShiftModifier)) {
+            score->cmdDeleteSelection();
+      }
```
dito, also unneeded

Jojo-Schmitz commented on mscore/pianoroll/pianoview.cpp:
```diff
@@ -1996,44 +2300,80 @@ void PianoView::pasteNotesAtCursor()
+      for (Note* note : notes) {
+            selection.add(note);
+            note->setSelected(true);
+      }
```
dito

Jojo-Schmitz commented on mscore/pianoroll/pianoview.cpp:
```diff
@@ -2056,6 +2396,10 @@ void PianoView::pasteNotes(const QString& copiedNotes, Fraction pasteStartTick,
+                        tickLen += lengthOffset;
+                        if (tickLen.numerator() <= 0) {
+                              continue;
+                        }
```
dito

Jojo-Schmitz commented on mscore/pianoroll/pianoview.cpp:
```diff
@@ -2109,26 +2455,122 @@ void PianoView::drawDraggedNotes(QPainter* painter)
+                int track = (int)_staff->idx() * VOICES + _editNoteVoice;
+
+                drawDraggedNote(painter, startTickFrac, endTickFrac - startTickFrac, pitch, track, _colorNoteDrag);
+          }
```
dito

Jojo-Schmitz commented on mscore/pianoroll/pianoview.cpp:
```diff
@@ -2109,26 +2455,122 @@ void PianoView::drawDraggedNotes(QPainter* painter)
+                drawDraggedNote(painter, startTickFrac, endTickFrac - startTickFrac, pitch, track, _colorNoteDrag);
+          }
+          return;
+      }
```
dito

Jojo-Schmitz commented on mscore/pianoroll/pianoview.cpp:
```diff
@@ -2149,6 +2591,10 @@ void PianoView::drawDraggedNotes(QPainter* painter)
+                        tickLen += pasteLengthOffset;
+                        if (tickLen.numerator() <= 0) {
+                            continue;
+                        }
```
dito

Jojo-Schmitz commented on mscore/pianoroll/pianoview.cpp:
```diff
@@ -510,6 +520,117 @@ void PianoView::drawBackground(QPainter* p, const QRectF& r)
+//---------------------------------------------------------
+
+void PianoView::drawNoteBlock(QPainter* p, PianoItem* block)
+{
```
still a brace indent issue

blackears replied:
Should be fixed now.

Jojo-Schmitz replied:
Not the one at the end of that method ;-)

worldwideweary commented:
> @worldwideweary: Testers wanted ;-)

Touché ;-)

Jojo-Schmitz commented:
Now that I fixed that typo ;-)
